File: Python/nrz_m.py
"""
Name : Vedang Kamat
Roll No : 22B-ET-074
Class : BE, SEM VII
Branch : Electronics and Telecommunication Engineering
Goa College of Engineering, Farmagudi
"""

import matplotlib.pyplot as plt

# ========== NRZ-M Encoding Function ==========
def nrzm_encode(bits):
    """NRZ-M encoding: 1 -> invert level, 0 -> keep previous"""
    encoded = []
    current = 1  # starting level (V)
    for b in bits:
        if b == 1:
            current = 0 if current == 1 else 1  # invert
        encoded.append(current)
    return encoded

# ========== Main ==========
def main():
    # Default bitstream
    default_bits = "100000000100001"

    # Input bit sequence (string)
    input_str = input("Enter the bit sequence [default: 100000000100001]: ").strip()
    if input_str == "":
        input_str = default_bits

    # Convert to integer list
    try:
        bits = [int(ch) for ch in input_str]
        if any(b not in (0,1) for b in bits):
            raise ValueError
    except ValueError:
        print("Invalid input! Only 0s and 1s allowed.")
        return

    n = len(bits)

    # Apply NRZ-M encoding
    encoded = nrzm_encode(bits)

    # Extend lists by 1 for proper plotting
    x = list(range(n + 1))
    bits_ext = bits + [bits[-1]]         
    encoded_ext = encoded + [encoded[-1]]  

    plt.figure(figsize=(12, 6))

    # ---- Input Bitstream ----
    plt.subplot(2,1,1)
    plt.step(x, bits_ext, where='post', color='black', linewidth=2)
    plt.fill_between(x, bits_ext, step='post', alpha=0.3, color='black')
    plt.ylim(-0.2, 1.2)
    plt.yticks([0,1], ["0","1"])
    plt.xticks(range(n+1))
    plt.xlim(0, n)
    plt.grid(True, linestyle='--', linewidth=0.7, color='gray')
    plt.title("Input Bitstream")

    # ---- NRZ-M Encoded Signal ----
    plt.subplot(2,1,2)
    plt.step(x, encoded_ext, where='post', color='black', linewidth=2)
    plt.fill_between(x, encoded_ext, step='post', alpha=0.3, color='black')
    plt.ylim(-0.2, 1.2)
    plt.yticks([0,1], ["0","V"])
    plt.xticks(range(n+1))
    plt.xlim(0, n)
    plt.grid(True, linestyle='--', linewidth=0.7, color='gray')
    plt.title("NRZ-M Encoded Signal")

    plt.xlabel("Bit Index")
    plt.tight_layout()
    plt.show()

if __name__ == "__main__":
    main()

File: C/nrz_m.c
/*
 Name : Vedang Kamat
 Roll No : 22B-ET-074
 Class : BE, SEM VII
 Branch : Electronics and Telecommunication Engineering
 Goa College of Engineering, Farmagudi
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Function for NRZ-M encoding
void nrzm_encode(int *b, int n) {
    printf("NRZ-M: ");
    int current = 1;
    for (int i = 0; i < n; i++) {
        if (b[i] == 1) current = !current;
        printf(current ? "+V | " : "0 | ");
    }
    printf("\n");
}

int main() {
    char *input;   // pointer for input string
    int *b;        // dynamic array for bits
    int n;         // number of bits
    size_t max_len = 100; // maximum input length (can adjust)

    // Allocate memory for input string
    input = (char *)malloc((max_len + 1) * sizeof(char)); // +1 for null terminator
    if (input == NULL) {
        printf("Memory allocation failed!\n");
        return 1;
    }

    // Input bit sequence as a string
    printf("Enter the bit sequence: ");
    scanf("%s", input);

    // Determine size automatically
    n = strlen(input);

    // Allocate memory for integer array
    b = (int *)malloc(n * sizeof(int));
    if (b == NULL) {
        printf("Memory allocation failed!\n");
        free(input);
        return 1;
    }

    // Convert string characters to integers
    for (int i = 0; i < n; i++) {
        if (input[i] == '0')
            b[i] = 0;
        else if (input[i] == '1')
            b[i] = 1;
        else {
            printf("Invalid character detected!\n");
            free(b);
            free(input);
            return 1;
        }
    }

    // Call NRZ-M encoding function
    nrzm_encode(b, n);

    // Free allocated memory
    free(b);
    free(input);
    return 0;
}

